src/transform/src/predicate_propagation.rs

```diff
@@ -0,0 +1,837 @@
+                    class.extend(
+                        classes
+                            .iter()
+                            .position(|c: &Vec<MirScalarExpr>| c.contains(&expr))
+                            .map(|p| classes.remove(p))
+                            .unwrap_or_else(|| vec![expr]),
+                    );
+                }
+                if class.len() > 1 {
+                    classes.push(class);
+                }
+            }
+            Constraint::Predicate(_) => {
+                predicate.reduce(input_type);
+                other_predicates.push(predicate);
+            }
+        }
+    }
+
+    *predicates = other_predicates;
+
+    'retry_loop: loop {
+        // Order each class so that literals come first, then column references, then weird things.
+        for class in classes.iter_mut() {
+            // Reduce the expressions within the class
+            for expr in class.iter_mut() {
+                expr.reduce(input_type);
+            }
+            class.sort_by(cmp_expr);
+            class.dedup();
+            // If we have a second literal, not equal to the first, we have a contradiction and can
+            // just replace everything with false.
+            if let Some(second) = class.get(1) {
+                if second.is_literal_ok() {
+                    predicates.clear();
+                    predicates.push(Constraint::Predicate(MirScalarExpr::literal_ok(
+                        Datum::False,
+                        ScalarType::Bool,
+                    )));
+                    return;
+                }
+            }
+        }
+        // Sort by complexity of representative, and perform substitutions within predicates.
+        classes.sort_by(|x, y| cmp_expr(&x[0], &y[0]));
+
+        let mut structured = PredicateStructure::default();
+        for class in classes.iter_mut() {
+            if class
+                .iter_mut()
+                .map(|expr| optimize(expr, &structured))
+                .fold(false, |acc, v| acc || v)
+            {
+                // Optimizing a member of the current class may result in a change in the
+                // order of the equivalente expression within the class and among the classes.
+                continue 'retry_loop;
+            }
+            let mut iter = class.iter();
+            if let Some(representative) = iter.next() {
+                for other in iter {
+                    structured.replacements.insert(other, representative);
+                }
+            }
+        }
+
+        // Visit each predicate and rewrite using the representative from each class.
+        for predicate in predicates.iter_mut() {
+            predicate.optimize(&structured);
+            // Simplify the predicate after the replacements
+            predicate.reduce(input_type);
+            // Add the optimized predicate to the knowledge base
+            structured.add_predicate(predicate);
+        }
```

Comment:
`expr::canonicalize::canonicalize_predicates` does a subset of what this method does, correct? I wonder if some of this code can be shared.

Reply:
Yeah, that is listed as a follow-up task.